chain-tx-enclave/mls/src/keypackage.rs

```diff
@@ -0,0 +1,158 @@
+//! Key package of mls protocol
+use parity_scale_codec::{Decode, Encode};
+use ring::rand::SystemRandom;
+use ring::signature::{self, EcdsaKeyPair, VerificationAlgorithm};
+#[cfg(feature = "edp")]
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use ra_client::{EnclaveCertVerifier, EnclaveCertVerifierConfig, EnclaveCertVerifierError};
+#[cfg(feature = "edp")]
+use ra_enclave::{EnclaveRaContext, EnclaveRaContextError};
+
+use crate::key::{PublicKey, SecretKey};
+
+pub type ProtocolVersion = u8;
+pub type Timespec = u64;
+
+pub const PROTOCOL_VERSION_MLS10: ProtocolVersion = 0;
+pub const DEFAULT_LIFE_TIME: Timespec = 30 * 24 * 3600;
+
+/// - Cipher suites fixed to MLS10_128_DHKEMP256_AES128GCM_SHA256_P256, don't support customization
+#[derive(Encode, Decode)]
+pub enum Extension {
+    SupportedVersions(ProtocolVersion),
+    Expiration(Timespec),
+    KeyID,
+    ParentHash,
+}
+
+#[derive(Encode, Decode)]
+pub struct KeyPackagePayload {
+    pub version: ProtocolVersion,
+    pub init_key: PublicKey,
+    pub credential: Vec<u8>, // Remote Attestation X509 certificate
+    pub extensions: Vec<Extension>,
+}
+
+impl KeyPackagePayload {
+    /// Sign payload with private key of credential
+    pub fn sign(&self, priv_key: &[u8]) -> Vec<u8> {
+        let msg = self.encode();
+        let key_pair =
+            EcdsaKeyPair::from_pkcs8(&signature::ECDSA_P256_SHA256_ASN1_SIGNING, priv_key).unwrap();
+        let system_random = SystemRandom::new();
+        let signature = key_pair.sign(&system_random, &msg).unwrap();
+        signature.as_ref().to_vec()
+    }
+
+    /// Verify key package payload
+    pub fn verify(
+        &self,
+        ra_config: EnclaveCertVerifierConfig,
+        now: Timespec,
+    ) -> Result<Vec<u8>, KeyPackageVerifyError> {
+        // Check for required extensions
+        if !self
+            .extensions
+            .iter()
+            .any(|ext| matches!(ext, Extension::SupportedVersions(_)))
+        {
+            return Err(KeyPackageVerifyError::ExtensionNotFound(
+                "SupportedVersions",
+            ));
+        }
+
+        let expire = self
+            .extensions
+            .iter()
+            .filter_map(|ext| match ext {
+                Extension::Expiration(expire) => Some(*expire),
+                _ => None,
+            })
+            .next()
+            .ok_or(KeyPackageVerifyError::ExtensionNotFound("Expiration"))?;
```

Comment:
clippy often recommends
ok_or_else

Reply:
`ok_or_else` is better when the construction of error is non trivial, I think `ok_or` is good here.